File: src/domain/tool-activity.ts
/**
 * Shared tool-use activity model for the CLI streaming renderer.
 *
 * The GUI (gui/src-tauri/src/lib.rs) and the TUI (tui/src/main.rs) already
 * ship an identical Rust implementation in `crates/harness-data/src/tool_activity.rs`.
 * This module is the TypeScript mirror — add new tool cases to both sides so
 * the three surfaces stay in sync (OSS-06 parity contract).
 */

/** Max entries retained in the activity stack. Matches the Rust/GUI cap. */
export const ACTIVITY_STACK_MAX = 20;

/** Default number of newest entries shown when the stack is collapsed. */
export const ACTIVITY_TOP_N = 3;

export interface ToolActivityEntry {
  /** Short human-readable description (e.g. "Reading foo.ts" or "$ pnpm test"). */
  text: string;
  /** Unix epoch milliseconds at the time the event arrived. */
  ts: number;
}

/**
 * Build a skimmable one-liner describing a Claude `tool_use` block. The
 * output format deliberately mirrors the Rust side — keep them aligned.
 */
export function describeToolUse(
  name: string,
  input: Record<string, unknown> | null | undefined
): string {
  const getStr = (key: string): string => {
    if (!input) return "";
    const v = (input as Record<string, unknown>)[key];
    return typeof v === "string" ? v : "";
  };
  const basename = (p: string): string => {
    const idx = p.lastIndexOf("/");
    return idx < 0 ? p : p.slice(idx + 1);
  };

  switch (name) {
    case "Read":
      return `Reading ${basename(getStr("file_path"))}`;
    case "Edit":
      return `Editing ${basename(getStr("file_path"))}`;
    case "Write":
      return `Writing ${basename(getStr("file_path"))}`;
    case "Bash":
      return `$ ${truncateChars(getStr("command"), 60)}`;
    case "Grep":
      return `Searching '${truncateChars(getStr("pattern"), 40)}'`;
    case "Glob":
      return `Finding ${getStr("pattern")}`;
    case "Agent": {
      const desc = getStr("description");
      return desc ? `Agent: ${desc}` : "Spawning agent";
    }
    case "WebSearch":
      return `Web search: ${truncateChars(getStr("query"), 40)}`;
    case "WebFetch":
      return `Fetching ${truncateChars(getStr("url"), 50)}`;
    case "LSP":
      return `LSP ${getStr("method")}`;
    case "Skill":
      return `/${getStr("skill")}`;
    default:
      return name;
  }
}

/**
 * Parse a single line of Claude `stream-json --verbose` stdout into an
 * activity description (or `null` if the line isn't a `tool_use` block we
 * want to visualise).
 *
 * Returns `null` for text chunks, result blocks, malformed JSON, or anything
 * else — callers should treat a null return as "skip, not an error".
 */
export function parseClaudeStreamLine(line: string): string | null {
  const trimmed = line.trim();
  if (!trimmed) return null;
  let json: unknown;
  try {
    json = JSON.parse(trimmed);
  } catch {
    return null;
  }
  if (!json || typeof json !== "object") return null;
  const obj = json as Record<string, unknown>;
  if (obj.type !== "assistant") return null;

  const message = obj.message;
  if (!message || typeof message !== "object") return null;
  const content = (message as { content?: unknown }).content;
  if (!Array.isArray(content)) return null;

  // Claude can emit multiple content blocks per assistant event. We return the
  // first tool_use block; if more than one tool_use appears in the same event
  // the caller can re-scan (rare in practice).
  for (const block of content) {
    if (!block || typeof block !== "object") continue;
    const b = block as Record<string, unknown>;
    if (b.type !== "tool_use") continue;
    const name = typeof b.name === "string" ? b.name : "tool";
    const inputRaw = b.input;
    const input =
      inputRaw && typeof inputRaw === "object"
        ? (inputRaw as Record<string, unknown>)
        : undefined;
    return describeToolUse(name, input);
  }
  return null;
}

/**
 * Append an entry to an existing stack, capping BEFORE the push so a burst of
 * tool_use events can't momentarily exceed the cap (gap #12 lesson — don't
 * append-then-trim).
 */
export function appendActivityCapped(
  stack: ToolActivityEntry[],
  entry: ToolActivityEntry,
  cap: number = ACTIVITY_STACK_MAX
): ToolActivityEntry[] {
  const next = stack.length >= cap ? stack.slice(stack.length - cap + 1) : stack.slice();
  next.push(entry);
  return next;
}

function truncateChars(s: string, n: number): string {
  // Use Array.from so surrogate-paired emoji count as a single unit instead of
  // being split mid-codepoint.
  const chars = Array.from(s);
  if (chars.length <= n) return s;
  return `${chars.slice(0, n).join("")}…`;
}

File: crates/harness-data/src/tool_activity.rs
//! Shared tool-use activity rendering for the GUI (src-tauri), the TUI, and
//! any other Rust surface that needs to turn a Claude stream-json `tool_use`
//! block into a one-line skimmable description.
//!
//! The CLI has a parallel TypeScript implementation in
//! `src/domain/tool-activity.ts` — keep the two in sync when adding cases.
//! These two sources are the cross-dashboard contract for tool-use
//! visualization parity (ticket OSS-06).

use serde::{Deserialize, Serialize};
use serde_json::Value;

/// Max number of activity entries to retain per stream. Matches the GUI's
/// `ACTIVITY_STACK_MAX` so cap-before-append behavior is identical across
/// surfaces.
pub const ACTIVITY_STACK_MAX: usize = 20;

/// Default number of newest activity entries to show when the stack is
/// collapsed. Matches the GUI's `ACTIVITY_TOP_N`.
pub const ACTIVITY_TOP_N: usize = 3;

/// One entry in a streaming activity stack. Matches the shape used by the
/// GUI's `ActivityEntry` and the TUI's `ActivityStack.entries`, serialized as
/// camelCase so it round-trips with the TS side if we ever persist it.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolActivityEntry {
    /// Human-readable one-liner (e.g. "Reading foo.ts" or "$ pnpm test").
    pub text: String,
    /// Unix epoch milliseconds. The TUI stores a formatted time string; this
    /// field is the canonical numeric form used by the GUI/CLI.
    pub ts: i64,
}

/// Build a human-readable one-liner describing what a `tool_use` block from
/// Claude's stream-json output is doing. Deliberately short — the goal is a
/// skimmable feed, not a full payload dump.
///
/// Keep this mirrored with `describeToolUse` in `src/domain/tool-activity.ts`
/// (the CLI/TS side). Adding a new case? Add it there too.
pub fn describe_tool_use(name: &str, input: &Value) -> String {
    let get_str = |k: &str| {
        input
            .get(k)
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string()
    };
    let basename = |p: &str| p.rsplit('/').next().unwrap_or(p).to_string();

    match name {
        "Read" => format!("Reading {}", basename(&get_str("file_path"))),
        "Edit" => format!("Editing {}", basename(&get_str("file_path"))),
        "Write" => format!("Writing {}", basename(&get_str("file_path"))),
        "Bash" => format!("$ {}", truncate_chars(&get_str("command"), 60)),
        "Grep" => format!("Searching '{}'", truncate_chars(&get_str("pattern"), 40)),
        "Glob" => format!("Finding {}", get_str("pattern")),
        "Agent" => {
            let desc = get_str("description");
            if desc.is_empty() {
                "Spawning agent".to_string()
            } else {
                format!("Agent: {}", desc)
            }
        }
        "WebSearch" => format!("Web search: {}", truncate_chars(&get_str("query"), 40)),
        "WebFetch" => format!("Fetching {}", truncate_chars(&get_str("url"), 50)),
        "LSP" => format!("LSP {}", get_str("method")),
        "Skill" => format!("/{}", get_str("skill")),
        _ => name.to_string(),
    }
}

/// Truncate a string to `n` chars (not bytes), appending `…` when clipped.
/// Unicode-safe: uses `chars().count()` / `chars().take()` so multibyte
/// sequences aren't split mid-codepoint.
fn truncate_chars(s: &str, n: usize) -> String {
    if s.chars().count() <= n {
        s.to_string()
    } else {
        let mut out: String = s.chars().take(n).collect();
        out.push('…');
        out
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn read_uses_basename() {
        let d = describe_tool_use("Read", &json!({ "file_path": "/abs/path/src/foo.ts" }));
        assert_eq!(d, "Reading foo.ts");
    }

    #[test]
    fn bash_truncates_long_commands() {
        let cmd = "a".repeat(200);
        let d = describe_tool_use("Bash", &json!({ "command": cmd }));
        // $ + 60 chars + ellipsis
        assert!(d.starts_with("$ "));
        assert!(d.ends_with('…'));
        assert_eq!(d.chars().count(), 2 + 60 + 1);
    }

    #[test]
    fn unknown_tool_falls_back_to_name() {
        let d = describe_tool_use("Mystery", &json!({}));
        assert_eq!(d, "Mystery");
    }

    #[test]
    fn skill_formats_with_slash() {
        let d = describe_tool_use("Skill", &json!({ "skill": "review-pr" }));
        assert_eq!(d, "/review-pr");
    }

    #[test]
    fn truncate_handles_multibyte() {
        // Smoke test: 100 ✓ chars, truncate to 5 — must not panic on UTF-8
        // boundary and must preserve full codepoints.
        let s = "✓".repeat(100);
        let t = truncate_chars(&s, 5);
        assert_eq!(t.chars().count(), 6); // 5 + ellipsis
    }

    #[test]
    fn activity_entry_roundtrips_camelcase() {
        let e = ToolActivityEntry {
            text: "Reading foo.ts".into(),
            ts: 1_700_000_000_000,
        };
        let s = serde_json::to_string(&e).unwrap();
        assert!(s.contains("\"text\""));
        assert!(s.contains("\"ts\""));
        let back: ToolActivityEntry = serde_json::from_str(&s).unwrap();
        assert_eq!(back.text, "Reading foo.ts");
    }
}
